provider/alibaba_cloud.go

```diff
@@ -0,0 +1,1079 @@
+	MaxChangeCount       int
+	EvaluateTargetHealth bool
+	AssumeRole           string
+	vpcID                string // Private Zone only
+	dryRun               bool
+	dnsClient            AlibabaCloudDNSAPI
+	pvtzClient           AlibabaCloudPrivateZoneAPI
+	privateZone          bool
+	clientLock           sync.RWMutex
+	nextExpire           time.Time
+}
+
+type alibabaCloudConfig struct {
+	RegionID        string    `json:"regionId" yaml:"regionId"`
+	AccessKeyID     string    `json:"accessKeyId" yaml:"accessKeyId"`
+	AccessKeySecret string    `json:"accessKeySecret" yaml:"accessKeySecret"`
+	VPCID           string    `json:"vpcId" yaml:"vpcId"`
+	RoleName        string    `json:"-" yaml:"-"` // For ECS RAM role only
+	StsToken        string    `json:"-" yaml:"-"`
+	ExpireTime      time.Time `json:"-" yaml:"-"`
+}
+
+// NewAlibabaCloudProvider creates a new Alibaba Cloud provider.
+//
+// Returns the provider or an error if a provider could not be created.
+func NewAlibabaCloudProvider(configFile string, domainFilter DomainFilter, zoneIDFileter ZoneIDFilter, zoneType string, dryRun bool) (*AlibabaCloudProvider, error) {
+	cfg := alibabaCloudConfig{}
+	if configFile != "" {
+		contents, err := ioutil.ReadFile(configFile)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to read Alibaba Cloud config file '%s': %v", configFile, err)
+		}
+		err = yaml.Unmarshal(contents, &cfg)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse Alibaba Cloud config file '%s': %v", configFile, err)
+		}
+	} else {
+		var tmpError error
+		cfg, tmpError = getCloudConfigFromStsToken()
+		if tmpError != nil {
+			return nil, fmt.Errorf("Failed to getCloudConfigFromStsToken: %v", tmpError)
+		}
+	}
+
+	// Public DNS service
+	var dnsClient AlibabaCloudDNSAPI
+	var err error
+
+	if cfg.RoleName == "" {
+		dnsClient, err = alidns.NewClientWithAccessKey(
+			cfg.RegionID,
+			cfg.AccessKeyID,
+			cfg.AccessKeySecret,
+		)
+	} else {
+		dnsClient, err = alidns.NewClientWithStsToken(
+			cfg.RegionID,
+			cfg.AccessKeyID,
+			cfg.AccessKeySecret,
+			cfg.StsToken,
+		)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Alibaba Cloud DNS client: %v", err)
+	}
+
+	// Private DNS service
+	var pvtzClient AlibabaCloudPrivateZoneAPI
+	if cfg.RoleName == "" {
+		pvtzClient, err = pvtz.NewClientWithAccessKey(
+			"cn-hangzhou", // The Private Zone location is fixed
+			cfg.AccessKeyID,
+			cfg.AccessKeySecret,
+		)
+	} else {
+		pvtzClient, err = pvtz.NewClientWithStsToken(
+			cfg.RegionID,
+			cfg.AccessKeyID,
+			cfg.AccessKeySecret,
+			cfg.StsToken,
+		)
+	}
+
+	provider := &AlibabaCloudProvider{
+		domainFilter: domainFilter,
+		zoneIDFilter: zoneIDFileter,
+		vpcID:        cfg.VPCID,
+		dryRun:       dryRun,
+		dnsClient:    dnsClient,
+		pvtzClient:   pvtzClient,
+		privateZone:  zoneType == "private",
+	}
+
+	if cfg.RoleName != "" {
+		provider.setNextExpire(cfg.ExpireTime)
+		go provider.refreshStsToken(1 * time.Second)
```

Comment:
maybe this could be also configurable via flag or something. refreshing every 1 second sounds eager to me. WDTY?

Reply:
- `roleName` is a Alibaba Cloud STS Token role, it can be fetch from ECS internal API, so it don't need to be config via flag.
- `refreshStsToken`, the first time is 1 second, but from the next time it will be turned to 10 minute.
```
time.Sleep(sleepTime)
		now := time.Now()
		sleepTime = p.nextExpire.Sub(nowTime)
		log.Infof("Distance expiration time %v", sleepTime)
		if sleepTime < time.Duration(10*time.Minute) {
			sleepTime = time.Duration(time.Second * 1)
		} else {
			sleepTime = time.Duration(9 * time.Minute)
                }
```
So in most situation, the refresh loop time will be 10 minute. I think it's fine.